docs/2_customization/custom_to_es.md

```diff
@@ -0,0 +1,159 @@
+  def read_file
+    YAML.load_file(@file_location)
+  end
+end
+```
+
+### subdocs
+
+The next thing you will need to address if your format needs to be split into
+multiple documents (such as personography files, spreadsheets, database dumps,
+etc), is how to split up a file. By default, Datura assumes your file is one
+item. If that is not the case, override `subdocs`:
+
+```
+def subdocs
+  Array(@file)
+end
+```
+
+Change that to something which will return an array of items. For example, from
+our YAML example, you might have:
+
+```
+def subdocs
+  @file["texts"]
+end
+```
+Or for an XML file:
+```
+def subdocs
+  @file.xpath("//grouping")
+end
+```
+
+### build_es_documents
+
+You're almost done with `file_custom.rb`. You just need to kick off a class
+that will handle the transformation per sub-document. For simplicity's sake, if
+this is a totally new format that Elasticsearch hasn't seen before, I recommend
+leaving this method alone. You can move onto the next step,
+[CustomToEs](#customtoes).
+
+If you want to try to piggyback off of an existing Datura class, then you may
+need to override this method. Instead of calling `CustomToEs.new()` in it, you
+would instead need to add a `require_relative` path at the top of the file to
+your new class, and then call `YournewclassToEs.new()` from `build_es_documents`.
+
+In your new class, you could presumably do something like
+
+```
+class YournewclassToEs < XmlToEs
+  # now you have access to XmlToEs helpers for xpaths, etc
+end
+```
+
+## CustomToEs
+
+The files in the [custom_to_es](/lib/datura/to_es/custom_to_es) directory and
+[custom_to_es.rb](/lib/datura/to_es/custom_to_es.rb) give you the basic
+structure you need to create your own version of these files. Since
+Datura has no way of knowing what format might come its way, the majority of the
+methods in `custom_to_es/fields.rb` are empty.
+
+The only thing you **MUST** override is `get_id`.
+
+Create a file in your overrides directory called `custom_to_es.rb` and add the
+following:
+
+```
+class CustomToEs
+
+  def get_id
+    # include code here that returns an id
+    # it could be the @filename(false) to get a filename without extension
+    # or it could be `@item["identifier"] to get the value of a column, etc
+
+    # you may want to prepend a collection abbreviation to your id, like
+    # "nei.#{some_value}"
+  end
+
+end
+```
+
+You can also add preprocessing or postprocess here by overriding `create_json`.
+
+It is expected that you will override most of the methods in `fields.rb`. For
+example, you might set a category like:
+
+```
+def category
+  # your code here, referencing @item if necessary
+end
+```
+
+One more note: due to how `CustomToEs` is created, it is expecting a subdoc
+and the original file. This is because it accommodates for something like a
+personography file, where you may want to deal with an individual person as
+`@item` but need to reference `@file` to get information about the repository
```

Comment:
Most of the x_to_es files use
`@file_location` as the fullpath `/var/local/www/data/collections/x/etc` so that things can read and write it
`@filename` is a quick grab for just `id.xml` or whatever the file name is

You can also get that by doing a `self.filename()` method call from anything inheriting from `FileType` because it has a method to get the filename with or without an extension.

In the case of `custom_to_es` I also included `@file` because I needed some general name for a variable that could hold any format (CSV, XML, YAML, JSON). That's because custom doesn't know what a user might be overriding the default "read file" behavior with.  I can rename `@file` to something else if that would be helpful?  like, `@document` or something?  I would not be excited about changing `@file_location` and `@filename` since they are used by most of the `x_to_es` classes although it wouldn't be impossible!

Reply:
Ah, okay. That makes sense. I think `@file` is fine for the placeholder as it's a safe assumption that most new formats will be a file of some form or another. Could you maybe document this summary here in the markdown file above all the various mentions of them? :pray: